lib/ddq-backend-mysql.js

```diff
@@ -0,0 +1,374 @@
+
+module.exports = (configValidation, crypto, EventEmitter, mysql, timers) => {
+    var cycleCounter, writeRecord;
+
+    /**
+     * Updates heartbeatDate of a specific record and calls the callback that is
+     * passed in.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     * @param {Function} callback
+     */
+    function heartbeat(ddqBackendInstance, recordId, callback) {
+        ddqBackendInstance.connection.query(`UPDATE ??
+            SET heartbeatDate = NOW()
+            WHERE hash = ?
+                AND isProcessing = true
+                AND owner = ?;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId,
+                ddqBackendInstance.owner
+            ], callback
+        );
+    }
+
+
+    /**
+     * Updates a specific record so that it can be requeued.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     * @param {Function} callback
+     */
+    function requeue(ddqBackendInstance, recordId, callback) {
+        ddqBackendInstance.connection.query(`UPDATE ??
+            SET owner = null,
+                isProcessing = false,
+                requeued = false
+            WHERE hash = ?;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId
+            ], callback
+        );
+    }
+
+
+    /**
+     * Deletes a specific record from the database. If unsuccessful, this will
+     * call the command to requeue the record.
+     *
+     * @param {Object} ddqBackendInstance
+     * @param {string} recordId
+     * @param {Function} callback
+     */
+    function remove(ddqBackendInstance, recordId, callback) {
+        ddqBackendInstance.connection.query(`DELETE FROM ??
+            WHERE hash = ?
+                AND requeued = false
+                AND owner = ?;`,
+            [
+                ddqBackendInstance.config.table,
+                recordId,
+                ddqBackendInstance.owner
+            ], (err) => {
+                if (err) {
+                    requeue(ddqBackendInstance, recordId, callback);
+                }
+
+                callback(err);
+            }
+        );
+    }
+
+
+    /**
+     * Polls for a random record with the minimum delay defined by the
+     * config.pollingDelayMs.
+     *
+     * @param {Object} ddqBackendInstance
+     */
+    function poll(ddqBackendInstance) {
+        /**
+         * Queries the database to get a random record. Emits that random record
+         * along with methods to handle it.
+         */
+        function checkNow() {
+            var wrappedMessage;
+
+            ddqBackendInstance.poller = null;
+
+            ddqBackendInstance.connection.query(`SELECT *
+                FROM ??
+                WHERE isProcessing = false
+                    AND topics IN (?)
+                    OR topics IS NULL
+                ORDER BY RAND()
+                LIMIT 1;`,
```

Comment:
This may perform poorly when there are 10,000 or 100,000 messages queued.  Care to generate that many messages and then performance test this query?  Here's an alternative based on (this post)[http://stackoverflow.com/questions/1823306/mysql-alternatives-to-order-by-rand#answer-25887651].

    SELECT *
    FROM ??
    WHERE isProcessing = false
    AND rand() < 4 / (
        SELECT count(*)
        FROM ??
        WHERE isProcessing = false
        AND topics IN (?)
        OR topics IS NULL
    )
    AND topics IN (?)
    OR topics IS NULL
    ORDER BY RAND()
    LIMIT 1

You'll have to work out the extra arguments, but this query should be able to be tested by hand using the mysql client in order to see if there is a benefit.  Considering that on busy services with 100,000 messages flying around, there will be many processes eating through the messages and the query would get executed MANY times.  We want a solution that can scale well.